--- src/corelib/global/qglobal.cpp	2014-04-10 14:37:11.000000000 -0400
+++ src/corelib/global/qglobal.cpp	2014-08-06 21:44:23.000000000 -0400
@@ -1185,6 +1185,7 @@
     \value MV_10_7     OS X 10.7
     \value MV_10_8     OS X 10.8
     \value MV_10_9     OS X 10.9
+    \value MV_10_10    OS X 10.10
     \value MV_Unknown  An unknown and currently unsupported platform
 
     \value MV_CHEETAH  Apple codename for MV_10_0
@@ -1197,6 +1198,7 @@
     \value MV_LION     Apple codename for MV_10_7
     \value MV_MOUNTAINLION Apple codename for MV_10_8
     \value MV_MAVERICKS    Apple codename for MV_10_9
+    \value MV_YOSEMITE     Apple codename for MV_10_10
 
     \sa WinVersion, SymbianVersion
 */
@@ -1674,8 +1676,9 @@
 {
 #if !defined(Q_OS_IOS)
     SInt32 gestalt_version;
-    if (Gestalt(gestaltSystemVersion, &gestalt_version) == noErr) {
-        return QSysInfo::MacVersion(((gestalt_version & 0x00F0) >> 4) + 2);
+    if (Gestalt(gestaltSystemVersionMinor, &gestalt_version) == noErr) {
+        // add 2 because OS X 10.0 is 0x02 in the enum
+        return QSysInfo::MacVersion(gestalt_version + 2);
     }
 #endif
     return QSysInfo::MV_Unknown;
--- src/corelib/global/qglobal.h	2014-04-10 14:37:11.000000000 -0400
+++ src/corelib/global-alt/qglobal.h	2014-08-06 21:42:10.000000000 -0400
@@ -304,31 +304,69 @@
 #endif
 
 #ifdef Q_OS_DARWIN
-#  ifdef MAC_OS_X_VERSION_MIN_REQUIRED
-#    undef MAC_OS_X_VERSION_MIN_REQUIRED
-#  endif
-#  define MAC_OS_X_VERSION_MIN_REQUIRED MAC_OS_X_VERSION_10_4
 #  include <AvailabilityMacros.h>
-#  if !defined(MAC_OS_X_VERSION_10_3)
-#     define MAC_OS_X_VERSION_10_3 MAC_OS_X_VERSION_10_2 + 1
+#
+#  // Availability.h was introduced with the OS X 10.6 SDK
+#  if (defined(__MAC_OS_X_VERSION_MAX_ALLOWED) && __MAC_OS_X_VERSION_MAX_ALLOWED >= 1060) || \
+      (defined(MAC_OS_X_VERSION_MAX_ALLOWED) && MAC_OS_X_VERSION_MAX_ALLOWED >= 1060)
+#    include <Availability.h>
+#  endif
+#
+#  ifdef Q_OS_MACX
+#    if !defined(__MAC_OS_X_VERSION_MIN_REQUIRED) || __MAC_OS_X_VERSION_MIN_REQUIRED < 1040
+#       undef __MAC_OS_X_VERSION_MIN_REQUIRED
+#       define __MAC_OS_X_VERSION_MIN_REQUIRED 1040
+#    endif
+#    if !defined(MAC_OS_X_VERSION_MIN_REQUIRED) || MAC_OS_X_VERSION_MIN_REQUIRED < 1040
+#       undef MAC_OS_X_VERSION_MIN_REQUIRED
+#       define MAC_OS_X_VERSION_MIN_REQUIRED 1040
+#    endif
+#  endif
+#
+#  // Numerical checks are preferred to named checks, but to be safe
+#  // we define the missing version names in case Qt uses them.
+#
+#  if !defined(__MAC_10_4)
+#       define __MAC_10_4 1040
+#  endif
+#  if !defined(__MAC_10_5)
+#       define __MAC_10_5 1050
+#  endif
+#  if !defined(__MAC_10_6)
+#       define __MAC_10_6 1060
+#  endif
+#  if !defined(__MAC_10_7)
+#       define __MAC_10_7 1070
+#  endif
+#  if !defined(__MAC_10_8)
+#       define __MAC_10_8 1080
+#  endif
+#  if !defined(__MAC_10_9)
+#       define __MAC_10_9 1090
+#  endif
+#  if !defined(__MAC_10_10)
+#       define __MAC_10_10 101000
 #  endif
 #  if !defined(MAC_OS_X_VERSION_10_4)
-#       define MAC_OS_X_VERSION_10_4 MAC_OS_X_VERSION_10_3 + 1
+#       define MAC_OS_X_VERSION_10_4 1040
 #  endif
 #  if !defined(MAC_OS_X_VERSION_10_5)
-#       define MAC_OS_X_VERSION_10_5 MAC_OS_X_VERSION_10_4 + 1
+#       define MAC_OS_X_VERSION_10_5 1050
 #  endif
 #  if !defined(MAC_OS_X_VERSION_10_6)
-#       define MAC_OS_X_VERSION_10_6 MAC_OS_X_VERSION_10_5 + 1
+#       define MAC_OS_X_VERSION_10_6 1060
 #  endif
 #  if !defined(MAC_OS_X_VERSION_10_7)
-#       define MAC_OS_X_VERSION_10_7 MAC_OS_X_VERSION_10_6 + 1
+#       define MAC_OS_X_VERSION_10_7 1070
 #  endif
 #  if !defined(MAC_OS_X_VERSION_10_8)
-#       define MAC_OS_X_VERSION_10_8 MAC_OS_X_VERSION_10_7 + 1
+#       define MAC_OS_X_VERSION_10_8 1080
 #  endif
 #  if !defined(MAC_OS_X_VERSION_10_9)
-#       define MAC_OS_X_VERSION_10_9 MAC_OS_X_VERSION_10_8 + 1
+#       define MAC_OS_X_VERSION_10_9 1090
+#  endif
+#  if !defined(MAC_OS_X_VERSION_10_10)
+#       define MAC_OS_X_VERSION_10_10 101000
 #  endif
 #endif
 
@@ -1630,6 +1668,7 @@
         MV_10_7 = 0x0009,
         MV_10_8 = 0x000A,
         MV_10_9 = 0x000B,
+        MV_10_10 = 0x000C,
 
         /* codenames */
         MV_CHEETAH = MV_10_0,
@@ -1641,7 +1680,8 @@
         MV_SNOWLEOPARD = MV_10_6,
         MV_LION = MV_10_7,
         MV_MOUNTAINLION = MV_10_8,
-        MV_MAVERICKS = MV_10_9
+        MV_MAVERICKS = MV_10_9,
+        MV_YOSEMITE = MV_10_10
     };
     static const MacVersion MacintoshVersion;
 #endif
