diff -urN qt-everywhere-opensource-src-4.8.7/src/gui/kernel/qeventdispatcher_mac.mm qt-everywhere-opensource-src-4.8.7.new/src/gui/kernel/qeventdispatcher_mac.mm
--- a/src/gui/kernel/qeventdispatcher_mac.mm	2014-12-18 06:48:36.000000000 -0500
+++ b/src/gui/kernel/qeventdispatcher_mac.mm	2015-01-04 23:35:52.000000000 -0500
@@ -608,6 +608,11 @@
                     // dispatcher (e.g to stop a native dialog). But that call wrongly stopped
                     // 'session' as well. As a result, we need to restart all internal sessions:
                     d->temporarilyStopAllModalSessions();
+
+                    // Clean up the modal session list, call endModalSession.
+                    if (d->cleanupModalSessionsNeeded) {
+                        d->cleanupModalSessions();
+                    }
                 }
             } else {
                 d->nsAppRunCalledByQt = true;
@@ -632,6 +637,11 @@
                         // dispatcher (e.g to stop a native dialog). But that call wrongly stopped
                         // 'session' as well. As a result, we need to restart all internal sessions:
                         d->temporarilyStopAllModalSessions();
+
+                        // Clean up the modal seesion list, call endModalSession.
+                        if (d->cleanupModalSessionsNeeded) {
+                            d->cleanupModalSessions();
+                        }
                     }
                     retVal = true;
                 } else do {
@@ -1044,7 +1054,7 @@
     }
 
 #ifdef QT_MAC_USE_COCOA
-    if (d->cleanupModalSessionsNeeded)
+    if (d->cleanupModalSessionsNeeded && d->currentExecIsNSAppRun)
         d->cleanupModalSessions();
 #endif
 
